### PR TITLE
Fix `QuantumCircuit.compose` on unusual types and registers

### LINKED ISSUE #6583: Flattened conditional circuits give incorrect results
### Information

- **Qiskit Terra version**:
- **Python version**:
- **Operating system**:

### What is the current behavior?

Flattening a conditional circuit gives incorrect results. It looks like the issue is probably in how the conditionals are implemented:

### Steps to reproduce the problem

Manually adding an additional registers to a teleport circuit works:
 
```python
def teleport_circuit():
    """Teleport qubit 0 -> 2"""
    qr = QuantumRegister(3)
    c0 = ClassicalRegister(1)
    c1 = ClassicalRegister(1)
    teleport = QuantumCircuit(qr, c0, c1)
    teleport.h(qr[1])
    teleport.cx(qr[1], qr[2])
    teleport.cx(qr[0], qr[1])
    teleport.h(qr[0])
    teleport.measure(qr[0], c0[0])
    teleport.measure(qr[1], c1[0])
    teleport.z(qr[2]).c_if(c0, 1)
    teleport.x(qr[2]).c_if(c1, 1)
    return teleport

# Add extra register for measuring qubit-2
c2 = ClassicalRegister(1)
circ = teleport_circuit()
circ.add_register(c2)
circ.measure(2, c2)

# Run
counts = AerSimulator().run(circ).result().get_counts(0)
marginal_counts(counts, [2])

# Marginal Counts (correct)
{'0': 1024}
```

But composing and flattening gives wrong answer:
```python
# Compose with larger circuit
circ = QuantumCircuit(3, 3)
circ = circ.compose(teleport_circuit(), range(3), range(2))
circ.measure(2, 2)

# Run
counts = AerSimulator().run(circ).result().get_counts(0)
marginal_counts(counts, [2])

# Marginal Counts (incorrect)
{'1': 228, '0': 796}
```

### What is the expected behavior?

Flattening the conditional circuit should implement the same circuit.

### Suggested solutions

From the drawer it looks liek the flattened circuit is handling conditions on the joint circuit wrong:
```
                    ┌───┐┌─┐              
q_0: ────────────■──┤ H ├┤M├──────────────
     ┌───┐     ┌─┴─┐└┬─┬┘└╥┘              
q_1: ┤ H ├──■──┤ X ├─┤M├──╫───────────────
     └───┘┌─┴─┐└───┘ └╥┘  ║  ┌───┐  ┌───┐ 
q_2: ─────┤ X ├───────╫───╫──┤ Z ├──┤ X ├─
          └───┘       ║   ║  └─╥─┘  └─╥─┘ 
                      ║   ║ ┌──╨──┐┌──╨──┐
c: 3/═════════════════╩═══╩═╡ = 1 ╞╡ = 2 ╞
                      1   0 └─────┘└─────┘
```
Looks like conditions are triggering on creg values `01` and `10`, but nothing on `11` which should trigger both conditional gates.

Looking at the qobj that gets assembled from these circuits we can see the difference:

Multi-register assembled qobj instructions:
`assemble(circ).experiments[0].instructions = `
```
[QasmQobjInstruction(name='h', qubits=[1]),
 QasmQobjInstruction(name='cx', qubits=[1, 2]),
 QasmQobjInstruction(name='cx', qubits=[0, 1]),
 QasmQobjInstruction(name='h', qubits=[0]),
 QasmQobjInstruction(name='measure', qubits=[0], register=[0], memory=[0]),
 QasmQobjInstruction(name='measure', qubits=[1], register=[1], memory=[1]),
 QasmQobjInstruction(name='bfunc', register=3, mask="0x1", relation="==", val="0x1"),
 QasmQobjInstruction(name='z', qubits=[2], conditional=3),
 QasmQobjInstruction(name='bfunc', register=4, mask="0x2", relation="==", val="0x2"),
 QasmQobjInstruction(name='x', qubits=[2], conditional=4),
 QasmQobjInstruction(name='measure', qubits=[2], register=[2], memory=[2])]
```

Flattened circuit assembled qobj instructions:
```
[QasmQobjInstruction(name='h', qubits=[1]),
 QasmQobjInstruction(name='cx', qubits=[1, 2]),
 QasmQobjInstruction(name='cx', qubits=[0, 1]),
 QasmQobjInstruction(name='h', qubits=[0]),
 QasmQobjInstruction(name='measure', qubits=[0], register=[0], memory=[0]),
 QasmQobjInstruction(name='measure', qubits=[1], register=[1], memory=[1]),
 QasmQobjInstruction(name='bfunc', register=3, mask="0x7", relation="==", val="0x1"),
 QasmQobjInstruction(name='z', qubits=[2], conditional=3),
 QasmQobjInstruction(name='bfunc', register=4, mask="0x7", relation="==", val="0x2"),
 QasmQobjInstruction(name='x', qubits=[2], conditional=4),
 QasmQobjInstruction(name='measure', qubits=[2], register=[2], memory=[2])]
```

You can see that in the second case the `mask` value is being set incorrectly to `0x7 = 0b111` in both case to the full 3-bit register rather than the the single bit values.

### LINKED ISSUE #6584: Cannot convert condition in circuit with multiple classical registers to instruction
### Information

- **Qiskit Terra version**:
- **Python version**:
- **Operating system**:

### What is the current behavior?

Appending a circuit with multiple conditional registers to an instruction raises an exception (from `to_instruction` method), however using `compose` instead of `append` seems to work.

A related issue is that the "working" circuit from compose assembles to incorrect qobj (see issue #6583)

### Steps to reproduce the problem

Take basic teleport circuit as example
```python
def teleport_circuit():
    """Teleport qubit_in to qubit_out"""
    # It would be nice to be able to do this without using
    # register objects

    # Teleport qubit 0 -> 2
    qr = QuantumRegister(3)
    c0 = ClassicalRegister(1)
    c1 = ClassicalRegister(1)
    teleport = QuantumCircuit(qr, c0, c1)
    teleport.h(qr[1])
    teleport.cx(qr[1], qr[2])
    teleport.cx(qr[0], qr[1])
    teleport.h(qr[0])
    teleport.measure(qr[0], c0[0])
    teleport.measure(qr[1], c1[0])
    teleport.z(qr[2]).c_if(c0, 1)
    teleport.x(qr[2]).c_if(c1, 1)
```
Appending this to another circuit with
```python
qc = QuantumCircuit(3, 2)
qc.append(teleport_circuit(), range(3), range(2))
```
raises
```
~/git/qiskit/qiskit-terra/qiskit/converters/circuit_to_instruction.py in circuit_to_instruction(circuit, parameter_map, equivalence_library)
    116             else:
    117                 raise QiskitError(
--> 118                     "Cannot convert condition in circuit with "
    119                     "multiple classical registers to instruction"
    120                 )

QiskitError: 'Cannot convert condition in circuit with multiple classical registers to instruction'
```
This error seems to come directly from `teleport_circuit().to_instruction()` which gets called by `append`.

If you do `compose` it produces a circuit, however the conditional values appear incorrect on the drawer and in the assembled qobj (see #6583)

```python
qc = QuantumCircuit(3, 3)
qc = qc.compose(teleport_circuit(), range(3), range(2))
print(qc.draw())
```
```
                    ┌───┐┌─┐              
q_0: ────────────■──┤ H ├┤M├──────────────
     ┌───┐     ┌─┴─┐└┬─┬┘└╥┘              
q_1: ┤ H ├──■──┤ X ├─┤M├──╫───────────────
     └───┘┌─┴─┐└───┘ └╥┘  ║  ┌───┐  ┌───┐ 
q_2: ─────┤ X ├───────╫───╫──┤ Z ├──┤ X ├─
          └───┘       ║   ║  └─╥─┘  └─╥─┘ 
                      ║   ║ ┌──╨──┐┌──╨──┐
c: 3/═════════════════╩═══╩═╡ = 1 ╞╡ = 2 ╞
                      1   0 └─────┘└─────┘
```

### What is the expected behavior?

`to_instruction` should work for any circuit by flatting registers.

### Suggested solutions

The implementation in compose seems to work, so maybe how it is implemented could be used for building the flattened circuit definition in the instruction-ified circuit.

### LINKED ISSUE #8691: `QuantumCircuit.compose` errors when `qubits` argument is `numpy.int64` instead of `int`.
### Environment

- `qiskit-terra: '0.21.0': 
- `python: '3.9.12'`: 
- Ubuntu-20.04 windows subsystem for Linux running on Windows 10: 
- `numpy: 1.21.5`

### What is happening?

`QuantumCircuit.compose()` errors when `qubits` argument is a list of `numpy.int64` instead of `int`.

### How can we reproduce the issue?

```python
from qiskit import QuantumCircuit
from qiskit.converters.circuit_to_dag import circuit_to_dag
import numpy as np

qc0 = QuantumCircuit(1)
qc1 = QuantumCircuit(1)
qc0.x(0)
qc1.x(0)

# First, let's try qubits = [0] (vanilla python `int`)
qc_vanilla_int = qc0.compose(qc1, qubits=[0])
# `QuantumCircuit.compose` does not error even when the `qubits` argument is invalid, so we need to check the DAG
circuit_to_dag(qc_vanilla_int) # does not error

# Now let's try qubits = [np.int64(0)]
qc_numpy_int = qc0.compose(qc1, qubits=[np.int64(0)]) #np.arange(0) does the same thing.
circuit_to_dag(qc_numpy_int) #this one errors
```

Here is the error:
```
DAGCircuitError: "(qu)bit 0 not found in OrderedDict([(Qubit(QuantumRegister(1, 'q'), 0), 
DAGOutNode(wire=Qubit(QuantumRegister(1, 'q'), 0)))])"
```

### What should happen?

`qc0.compose` should return a `QuantumCircuit` object that can be converted to a DAG.

### Any suggestions?

_No response_

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

`QuantumCircuit.compose` previously had its own handling for converting bit specifiers into bit instances, meaning its functionality was often surprisingly less permissive than `append`, or just incorrect.  This swaps the method to use the standard conversion utilities.

The logic for handling mapping conditional registers was previously the old logic in `DAGCircuit`, which was called without respect to re-ordering of the clbits during the composition.  The existing test case surrounding this that is modified by this commit made an incorrect assertion; in general, the condition was not the same after the composition.  In this particular test case it was ok because both bits were being tested for the same value, but had the condition value been 0b01 or 0b10 it would have been incorrect.

This commit updates the register-mapping logic to respect the reordering of clbits, and to create a new aliasing register to ensure the condition is represented exactly, if this is required.  This fixes a category of bug where too-small registers could incorrectly be mapped to larger registers.  This was not a valid transformation, because it created assertions about bits that were previously not involved in the condition's comparison.

### Details and comments

Fix #6583
Fix #6584
Fix #8691

Note that with this change in place, the circuit visualisation tools may well not display the circuits correctly; they don't handle aliasing registers well at the moment.  Similarly, I don't know precisely how well these objects will transport through the legacy Qobj paths; the circuits being described here may well not be representable in the Qobj model.  If that is the case, though, it'll be up to `assemble` to raise a suitable error; the representation here is what is correct for `QuantumCircuit`.

### CODE REVIEW

File: qiskit/circuit/quantumcircuit.py [hunk 1]
```diff
@@ -840,98 +840,100 @@ def compose(
...
-                qubits = list(range(other.num_qubits))
-
+                qubits = self.qubits[: other.num_qubits]
             if clbits is None:
-                clbits = list(range(other.num_clbits))
-
+                clbits = self.clbits[: other.num_clbits]
             if front:
-                dest.data.insert(0, CircuitInstruction(other, qubits, clbits))
+                old_data = list(dest.data)
```
Comment (Cryoris) [inline qiskit/circuit/quantumcircuit.py:866]: Why is rebuilding the list here required? Does the insertion at front create a bug? 🤔

File: test/python/circuit/test_compose.py [hunk 2]
```diff
@@ -411,16 +437,47 @@ def test_compose_conditional(self):
...
+        x = next(ins.operation for ins in test.data[::-1] if ins.operation.name == "x")
+        # The registers should have been mapped, including the bits inside them.  Unlike the
+        # previous test, there are no matching registers in the destination circuit, so the
+        # composition needs to add new registers (bit groupings) over the existing mapped bits.
+        self.assertIsNot(z.condition, None)
+        self.assertIsInstance(z.condition[0], ClassicalRegister)
+        self.assertEqual(len(z.condition[0]), len(right.cregs[0]))
+        self.assertIs(z.condition[0][0], test.clbits[0])
+        self.assertEqual(z.condition[1], 1)
+        #
```
Comment (Cryoris) [inline test/python/circuit/test_compose.py:474]:
```suggestion

```

**Review by Cryoris — COMMENTED** [review]
One small question, otherwise LGTM!

File: qiskit/circuit/quantumcircuit.py (hunk 1 above)
Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py:869]: The next line does `dest.clear`, which mutates `dest.data` in place, so we'd lose access to the data if we don't clone the references.  I'd thought there was a comment, but I guess I was thinking of somewhere else in Terra that does a similar thing.  Added a comment in a8dcfb5.

Comment (jakelishman) [reply qiskit/circuit/quantumcircuit.py:869]: Julien mentioned on Slack that his question was about whether `data.insert(0, ...)` caused a bug, or if this was a stylistic change.  The `insert(0, x)` form is incorrect: it invalidates all existing references in the `ParameterTable`.

Comment (Cryoris) [reply qiskit/circuit/quantumcircuit.py:869]: Thank you!

**Pull request merged**